doc/api/n-api.md

```diff
@@ -2914,6 +2944,36 @@ The native string is copied.
 The JavaScript `string` type is described in
 [Section 6.1.4][] of the ECMAScript Language Specification.
 
+#### `node_api_create_external_string_utf16`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
+
+```c
+napi_status node_api_create_string_utf16(napi_env env,
+                                         const char16_t* str,
+                                         size_t length,
+                                         napi_value* result)
+```
+
+* `[in] env`: The environment that the API is invoked under.
+* `[in] str`: Character buffer representing a UTF16-LE-encoded string.
+* `[in] length`: The length of the string in two-byte code units, or
+  `NAPI_AUTO_LENGTH` if it is null-terminated.
+* `[out] result`: A `napi_value` representing a JavaScript `string`.
+
+Returns `napi_ok` if the API succeeded.
+
+This API creates a JavaScript `string` value from a UTF16-LE-encoded C string.
+The native string may not be copied and must thus exist for the entire life
+cycle of the JavaScript value.
```

Comment:
https://github.com/nodejs/node/issues/48311

We could use `napi_add_finalizer()`.

Reply:
... or not. I forgot that references to primitive types cannot be weak. In that case, it makes sense to support a finalizer here, because we have the opportunity via `ExternalStringResource`'s `Dispose` method.

Reply:
Actually, it would be difficult to cleanly support this, because we create normal strings sometimes and external ones at other times. Specificially, we create normal strings if compiled with `V8_ENABLE_SANDBOX` and for UTF-8-encoded strings, for which V8 doesn't have an external string creation API. So, if the user supplied a finalizer and we ended up creating a regular string, we would have no way of knowing when it was garbage-collected.

Reply:
We could change the implementation from best-effort to guaranteed-or-error, meaning that, in the case of `V8_ENABLE_SANDBOX` `node_api_create_external_string_latin1()` and `node_api_create_external_string_utf16()` would return a new status `napi_no_external_strings_allowed`. `node_api_create_external_string_utf8()` would always return this status. Then, it would be up to add-on maintainers to provide a fallback mechanism.

Reply:
... or we could return a new status of `napi_string_copied` instead of `napi_ok` to indicate that the add-on can dispose of the native string.

Reply:
I believe that adding a new status could be a good option. Another option we could consider might be manually calling a deletor (if exists) just before returning `napi_ok` if the external string is copied.

Reply:
@mhdawson suggested we pass in a boolean as an out-parameter to record whether the string was copied or not. So, the API would look like this:

```c
napi_status node_api_create_external_string_*(napi_env env,
                                              const char* string,
                                              size_t length,
                                              napi_value* result,
                                              napi_finalize finalizer /* may be NULL */,
                                              void* hint              /* may be NULL */,
                                              bool* was_copied        /* may be NULL */);
```